### PR TITLE
fix: fail over and cool down on 401 token-invalidated responses (#171)

### LINKED ISSUE #171: Auth token invalidated error does not fail over to next healthy account
## Bug Description

When one stored OAuth account returns:

```text
Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

`oc-codex-multi-auth` appears to keep selecting that account for the requested model family instead of cooling it down/removing it or failing over to another valid stored account.

In this case, re-running `opencode auth login` did not fix the normal plugin path because the persisted family routing still pointed at the bad account slot. Manually switching `activeIndex` and every `activeIndexByFamily` entry from account slot `0` to account slot `1` immediately restored normal operation.

## Steps to Reproduce

1. Configure `oc-codex-multi-auth` with two OpenAI OAuth accounts.
2. Have account slot `0` selected in `~/.opencode/oc-codex-multi-auth-accounts.json`:

```json
{
  "activeIndex": 0,
  "activeIndexByFamily": {
    "gpt-5.4-mini": 0,
    "gpt-5.4": 0,
    "codex": 0
  }
}
```

3. Account slot `0` starts returning the OpenAI/Codex auth error:

```text
Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

4. Run a normal request through the plugin:

```bash
opencode run -m openai/gpt-5.4-mini "Reply exactly: OK"
```

5. Compare with `--pure`, which bypasses the plugin:

```bash
opencode --pure run -m openai/gpt-5.4-mini "Reply exactly: OK"
opencode --pure run -m openai/gpt-5.5-fast "Reply exactly: OK"
```

## Expected Behavior

When a selected account returns this token-invalidated auth error, the plugin should treat it as an auth/account health failure and either:

- refresh/re-authenticate that account if possible,
- mark the account or refresh-token group as cooling down/unhealthy,
- remove it after the configured auth failure threshold,
- or at least fail over to the next healthy account in the account pool for that request/model family.

`codex-health` / `codex-doctor --fix` should also be able to surface or repair this state without requiring manual JSON edits.

## Actual Behavior

Normal plugin mode failed repeatedly with:

```text
Error: Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

The failure persisted after re-login because the plugin continued routing the model family to account slot `0`.

At the same time, bypassing the plugin worked:

```text
opencode --pure run -m openai/gpt-5.4-mini "Reply exactly: OK"
# OK

opencode --pure run -m openai/gpt-5.5-fast "Reply exactly: OK"
# OK
```

After manually changing the multi-account routing to account slot `1`, normal plugin mode worked again:

```text
opencode run -m openai/gpt-5.4-mini "Reply exactly: OK"
# OK

opencode run -m openai/gpt-5.5-fast "Reply exactly: OK"
# OK
```

## Environment

- opencode version: `1.17.3`
- oc-codex-multi-auth version: `6.3.2`
- OS: macOS 26.2
- Node.js: `v26.0.0`
- Account pool: 2 OpenAI OAuth accounts

## Notes / Possible Cause

This looks related to the request/stream error path rather than the token refresh path:

- the plugin already has auth refresh failure handling, cooldown, and removal thresholds;
- rate limits, server errors, unsupported models, and deactivated workspaces appear to have rotation paths;
- but this token-invalidated `AI_APICallError` appears to bubble out as a normal request failure and does not mark the selected account unhealthy or try the next account.

Related but not identical to #156. That issue covered duplicate accounts / disabled valid logins. This case had two stored accounts and was fixed by moving family routing from the invalidated slot to the healthy slot.

## Compliance Checklist

- [x] I'm using this plugin for personal development only
- [x] I have an active ChatGPT Plus/Pro subscription
- [x] This issue is not related to attempting commercial use or TOS violations
- [x] I've reviewed the FAQ and Troubleshooting sections

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Fixes #171 — a stored OAuth account whose access token is invalidated server-side returns HTTP 401 (`Your authentication token has been invalidated. Please try signing in again.`), but the request pipeline had **no 401 handler**. The error fell straight through to `return errorResponse` without rotating or marking the account unhealthy, so persisted family routing kept pinning every request to the dead account slot. The only workaround was hand-editing `activeIndex`/`activeIndexByFamily` in `oc-codex-multi-auth-accounts.json`.

## Root cause

In the request loop (`index.ts`), non-OK responses are classified for workspace-deactivation, unsupported-model, recoverable, 5xx server, and rate-limit handling — each with a rotation/cooldown path. A 401 matched **none** of these and hit the generic fall-through (`lastErrorCategory = "http"; return errorResponse`), which neither cools the account down nor fails over. The plugin already refreshes tokens proactively, but a token invalidated *server-side* can still look unexpired locally, so no refresh fires and the stale token is rejected mid-request.

## Fix

Treat a request-path 401 as an account-health failure, mirroring the existing token-refresh failure handling:

- **New detectors:** `isInvalidatedAuthTokenError(errorBody, status)` (request-path, driven by HTTP 401 with structured-code/message fallbacks) and `isInvalidatedAuthTokenMessage()` (pure message matcher, single source of truth for the wording + appended `opencode auth login` hint).
- **Failover in the request loop:** refund the consumed token, record the failure, increment the shared auth-failure counter, cool the refresh-token group down, and rotate to the next healthy account. Past `MAX_AUTH_FAILURES_BEFORE_REMOVAL` the group is removed and traversal restarts. 401s intentionally **do not** feed the circuit breaker (it guards upstream faults, not client-side auth decisions).
- **Recovery:** clear the auth-failure counter on a successful request so a recovered account does not accumulate stale failures toward removal.
- **codex-health / codex-doctor:** the quota probe now flags `token-invalid` on an invalidated-token error, so `codex-doctor --fix` repairs the active routing without manual JSON edits.

Once the bad slot is cooled down, the next successful rotation persists the updated family routing, so the failure **self-heals across restarts** — no more manual `activeIndex` editing.

## Tests

- Unit tests for both detectors (`test/error-sentinels.test.ts`, `test/fetch-helpers.test.ts`), including non-matching rate-limit / entitlement / server-error cases.
- End-to-end regression test (`test/index-retry.test.ts`) asserting a 401 on the first account fails over to the second and returns 200.
- Full suite green: **2494 tests pass**, `tsc --noEmit` and `eslint` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PwrFRje1b1FmpjPE8grMPn

---
_Generated by [Claude Code](https://claude.ai/code/session_01PwrFRje1b1FmpjPE8grMPn)_

## Summary by CodeRabbit

* **New Features**
  * Automatic detection and rotation when authentication tokens are invalidated or revoked
  * Auth failure tracking with cooldown/removal of affected accounts after a failure threshold
* **Bug Fixes**
  * 401 “invalidated token” responses now trigger auth-failure handling instead of generic HTTP errors
  * Health-check/doctor probing now flags invalidated tokens for repair and removes them from rotation
  * Auth-failure state is cleared after successful requests
* **Tests**
  * Added coverage for invalidated-token detection and retry/rotation regression scenarios

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

this pr fixes issue #171 by adding an explicit 401 token-invalidated handler in the main request loop and the `codex-health`/`codex-doctor` probe path, replacing a silent fall-through that kept dead accounts permanently pinned in family routing.

- **request loop**: on a 401, the plugin refunds the token, increments an auth-failure counter, cools the refresh-token group down (or removes it past `MAX_AUTH_FAILURES_BEFORE_REMOVAL`), and rotates to the next healthy account; `clearAuthFailures` is called on the first success to prevent stale counts from accumulating.
- **detectors**: `isInvalidatedAuthTokenError` (primary signal is http 401; code/message fallbacks are narrow to avoid false-positives on the status-less probe path) and `isInvalidatedAuthTokenMessage` (single-source regex, deliberately requires an explicit token keyword to avoid matching generic session-expiry bodies).
- **codex-doctor**: the health probe normalises any 401 to the canonical invalidated-token message so `--fix` can flag and repair the dead routing slot without manual json edits.

<h3>Confidence Score: 5/5</h3>

safe to merge — the 401 failover path is correct, detectors are narrow and well-tested, and the success-path counter reset prevents stale accumulation

the core fix is sound: any 401 in the request loop now cools the account down and rotates instead of pinning the dead slot. detector logic is narrow and backed by thorough unit tests. the only gap is the `removedCount == 0` branch where the failure counter isn't cleared after a failed removal, causing repeated warn+cooldown cycles for standalone accounts — noisy and potentially chatty on disk, but rotation continues correctly and no data is lost

index.ts around the `logWarn` branch at `MAX_AUTH_FAILURES_BEFORE_REMOVAL` and the corresponding mock in test/index-retry.test.ts

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| index.ts | adds 401 token-invalidated handler in the request loop (cooldown + rotate) and the codex-health probe path; clears auth-failure counter on success; logic is sound but the `removedCount == 0` branch leaves failure counter elevated across cycles |
| lib/request/fetch-helpers.ts | adds `getStructuredErrorMessage` helper and `isInvalidatedAuthTokenError`; HTTP 401 status is the primary signal; code/message fallbacks are narrow and exclude generic permission codes |
| lib/error-sentinels.ts | adds `INVALIDATED_AUTH_TOKEN_MESSAGE_PATTERN` and `isInvalidatedAuthTokenMessage`; pattern is deliberately narrow to avoid false positives on the status-less probe path; well-documented |
| test/index-retry.test.ts | adds e2e failover test for 401 rotation; `removeAccountsWithSameRefreshToken` mock always returns 1, leaving the `removedCount == 0` branch (standalone account at threshold) uncovered |
| test/fetch-helpers.test.ts | covers 401-by-status, code fallbacks, message fallback, and non-matching cases; tests are accurate against the implementation |
| test/error-sentinels.test.ts | good coverage of the message pattern matcher including positive, negative, and edge cases; no issues found |
| test/index.test.ts | adds `isInvalidatedAuthTokenError` mock to the existing suite mock so the new handler doesn't break existing tests; minimal change, no issues |

</details>

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[fetch request] --> B{response.ok?}
    B -- yes --> C[recordSuccess\nclearAuthFailures\ncircuitBreaker.recordSuccess]
    B -- no --> D[parse errorBody]
    D --> E{deactivated\nworkspace?}
    E -- yes --> F[throw createDeactivatedWorkspaceError]
    E -- no --> G{rate limit\n429?}
    G -- yes --> H[cooldown + rotate]
    G -- no --> I{unsupported\nmodel?}
    I -- yes --> J[fallback model + rotate]
    I -- no --> K{isInvalidatedAuthTokenError\nstatus=401?}
    K -- yes --> L[refundToken\nrecordFailure\nincrementAuthFailures]
    L --> M{failures >=\nMAX_AUTH_FAILURES\n_BEFORE_REMOVAL?}
    M -- yes --> N[removeAccountsWithSameRefreshToken]
    N --> O{removedCount > 0?}
    O -- yes --> P[saveToDisk\ntoast\nattempted.clear\nbreak]
    O -- no --> Q[logWarn\n counter not cleared]
    Q --> R[markAccountsWithRefreshToken\nCoolingDown]
    M -- no --> R
    R --> S{cooledCount <= 0?}
    S -- yes --> T[markAccountCoolingDown]
    S -- no --> U[saveToDisk\nlogWarn\nmaybe toast\nbreak]
    T --> U
    K -- no --> V[generic HTTP error\nreturn errorResponse]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[fetch request] --> B{response.ok?}
    B -- yes --> C[recordSuccess\nclearAuthFailures\ncircuitBreaker.recordSuccess]
    B -- no --> D[parse errorBody]
    D --> E{deactivated\nworkspace?}
    E -- yes --> F[throw createDeactivatedWorkspaceError]
    E -- no --> G{rate limit\n429?}
    G -- yes --> H[cooldown + rotate]
    G -- no --> I{unsupported\nmodel?}
    I -- yes --> J[fallback model + rotate]
    I -- no --> K{isInvalidatedAuthTokenError\nstatus=401?}
    K -- yes --> L[refundToken\nrecordFailure\nincrementAuthFailures]
    L --> M{failures >=\nMAX_AUTH_FAILURES\n_BEFORE_REMOVAL?}
    M -- yes --> N[removeAccountsWithSameRefreshToken]
    N --> O{removedCount > 0?}
    O -- yes --> P[saveToDisk\ntoast\nattempted.clear\nbreak]
    O -- no --> Q[logWarn\n counter not cleared]
    Q --> R[markAccountsWithRefreshToken\nCoolingDown]
    M -- no --> R
    R --> S{cooledCount <= 0?}
    S -- yes --> T[markAccountCoolingDown]
    S -- no --> U[saveToDisk\nlogWarn\nmaybe toast\nbreak]
    T --> U
    K -- no --> V[generic HTTP error\nreturn errorResponse]
```

</a>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `test/index-retry.test.ts`, line 97-122 ([link](https://github.com/ndycode/oc-codex-multi-auth/blob/cb845b5eb2fe75d2c562fa1bd42a6d8bf5170de5/test/index-retry.test.ts#L97-L122)) 

   <a href="#">P2</a> **`markAccountCoolingDown` fallback path has no test coverage**

   the mock always returns `1` from `markAccountsWithRefreshTokenCoolingDown`, so the `cooledCount <= 0` branch that calls `markAccountCoolingDown` (index.ts ~line 2690) is never exercised. this fallback fires when no accounts share the refresh token (e.g. a standalone account not part of a token group), and on windows that's also the path that triggers `saveToDiskDebounced` writes on a single-account install — worth covering with a test that returns 0 from the group-cooldown mock to verify the per-account fallback fires instead.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: test/index-retry.test.ts
   Line: 97-122

   Comment:
   **`markAccountCoolingDown` fallback path has no test coverage**

   the mock always returns `1` from `markAccountsWithRefreshTokenCoolingDown`, so the `cooledCount <= 0` branch that calls `markAccountCoolingDown` (index.ts ~line 2690) is never exercised. this fallback fires when no accounts share the refresh token (e.g. a standalone account not part of a token group), and on windows that's also the path that triggers `saveToDiskDebounced` writes on a single-account install — worth covering with a test that returns 0 from the group-cooldown mock to verify the per-account fallback fires instead.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22ndycode%2Foc-codex-multi-auth%22%20on%20the%20existing%20branch%20%22claude%2Fgithub-issue-171-6kyby4%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Fgithub-issue-171-6kyby4%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20test%2Findex-retry.test.ts%0ALine%3A%2097-122%0A%0AComment%3A%0A**%60markAccountCoolingDown%60%20fallback%20path%20has%20no%20test%20coverage**%0A%0Athe%20mock%20always%20returns%20%601%60%20from%20%60markAccountsWithRefreshTokenCoolingDown%60%2C%20so%20the%20%60cooledCount%20%3C%3D%200%60%20branch%20that%20calls%20%60markAccountCoolingDown%60%20%28index.ts%20~line%202690%29%20is%20never%20exercised.%20this%20fallback%20fires%20when%20no%20accounts%20share%20the%20refresh%20token%20%28e.g.%20a%20standalone%20account%20not%20part%20of%20a%20token%20group%29%2C%20and%20on%20windows%20that's%20also%20the%20path%20that%20triggers%20%60saveToDiskDebounced%60%20writes%20on%20a%20single-account%20install%20%E2%80%94%20worth%20covering%20with%20a%20test%20that%20returns%200%20from%20the%20group-cooldown%20mock%20to%20verify%20the%20per-account%20fallback%20fires%20instead.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=ndycode%2Foc-codex-multi-auth&pr=172&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<sub>Reviews (2): Last reviewed commit: ["fix: tighten invalidated-token detectors..."](https://github.com/ndycode/oc-codex-multi-auth/commit/855102672a47fcb1b9e422fcf556f6421ddc7c30) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38059686)</sub>

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -2629,6 +2631,89 @@ while (attempted.size < Math.max(1, accountCount)) {
...
+																														accountManager.recordFailure(account, modelFamily, model);
+																														account.lastSwitchReason = "rotation";
+																														runtimeMetrics.failedRequests++;
+																														runtimeMetrics.accountRotations++;
+																														runtimeMetrics.lastError = `Auth token invalidated on ${accountLabel}`;
+																														runtimeMetrics.lastErrorCategory = "auth-invalidated";
+
+																														const failures = await accountManager.incrementAuthFailures(account);
+																														if (failures >= ACCOUNT_LIMITS.MAX_AUTH_FAILURES_BEFORE_REMOVAL) {
+																															const removedCount =
```
[the remark this answers is not in the document]
Comment (ndycode) [reply index.ts:2660]: Thanks — I looked at this but am intentionally leaving it as-is, for two reasons:

1. **`incrementAuthFailures` is already serialized.** It's not a bare read-modify-write: `lib/accounts/recovery.ts` chains every call for a given refresh token through a per-token promise (`incrementAuthFailuresChain`), so concurrent callers each `await` a distinct, monotonically increasing value (…→2→3→…) rather than both reading the same stale count. So the "both see a value below threshold and both cool down" case can't happen — exactly one caller observes the value that crosses `MAX_AUTH_FAILURES_BEFORE_REMOVAL`.

2. **The double-removal edge is already handled and benign.** If the counter is at the threshold and two requests cross it, the second `removeAccountsWithSameRefreshToken` returns 0 and we fall into the existing `logWarn` + `markAccountsWithRefreshTokenCoolingDown` fallback (which also no-ops gracefully on an already-removed group). No corruption, no duplicate state — just one extra warning line.

This deliberately mirrors the established token-refresh-failure path a few lines up (`const failures = await accountManager.incrementAuthFailures(account); if (failures >= MAX) { … }`), so the new 401 branch stays consistent with the accepted pattern rather than introducing a one-off mutex. Happy to revisit if you'd prefer to harden both paths together in a follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_
